deeplink/src/main/java/com/hellofresh/deeplink/BaseRoute.kt

```diff
@@ -78,4 +100,17 @@ abstract class BaseRoute<out T>(private vararg val routes: String) : Action<T> {
         return routes.contentHashCode()
     }
 
+    companion object {
+
+        /**
+         * Regex pattern that will be used to validate and retrieve path values from a specified input.
+         *
+         * This pattern is made up of 3 groups.
+         * Group 1: An optional parameter name that will be used to reference the actual path value
+         * Group 2: The entirety of the (optional) regex pattern for the path
+         * Group 3: The actual regex pattern that will be supplied by the user
+         */
+        private val PARAMETERIZED_PATH_PATTERN = Pattern.compile("^:(\\w*)(\\((.*)\\))?$")
```

Comment:
> What's making this difficult with the current syntax? Like trying to deduce grouping with the current use of () symbols ? if so will it make sense to replace that with {}?

I'm answering this here, so that we can reply directly to the comment, instead of having independent comments on the PR
The pattern we implemented with this PR is `:parameterName(REGEX)`, and a group could be defined like `:parameterName(XXX(group)YYY)`
As you can see, we only have 1 **parameterName**, so if one were to do something like `:parameterName(XXX(group1)YYY(group2)ZZZ)`, then we'll have to decide which of the groups we want to associate with `parameterName`, or just ignore them altogether. This consequently makes it quite difficult to support multiple groups seamlessly

A possible syntax could be something like this: `XXX:parameter1(group1)YYY:parameter2(group2)ZZZ`
This way, each group is directly matched to its own key, and the grouping problem is then solved.
This obviously comes with its own set of complexities, especially in supporting such syntax and how much performance impact it might bring
We could think along this direction anyway, if we will support grouping, or like I said, we can just skip grouping altogether in this first iteration.

--
Possible use case for single group is the recipe ID example I gave earlier.
I don't think we have any use case for multiple groups at the moment. Yet, once we support grouping, it might then feel incomplete that the user is only restricted to a single group.

Reply:
> A possible syntax could be something like this: `XXX:parameter1(group1)YYY:parameter2(group2)ZZZ`
> This way, each group is directly matched to its own key, and the grouping problem is then solved.

I see what you mean now. Using grouping for a single parameter posses the question when matched, which group values to use. If I understand that correctly that is. What happens if we treat its value as `hashMapOf<String, List<String>>` this will require us to change `val params = hashMapOf<String, String>()` to ` val params = hashMapOf<String, Any>` and might pose compatibility issues with previous usage. If this is too of a bad implementation, we can ignore it for now.

> This obviously comes with its own set of complexities, especially in supporting such syntax and how much performance impact it might bring
> We could think along this direction anyway, if we will support grouping, or like I said, we can just skip grouping altogether in this first iteration.

Yes if we don't have immediate use for it for now, let's drop its support and think through for a proper implementation. I think I get it much now.

Reply:
I thought about this possibility. `Map<String, List<String>>`

However, I dropped it because it fails in various cases:
1. It's backward incompatible in a big way. Haha. We could work our way around this though, and still sort of retain backward compatibility, but then we’ll have to carry the artefact along. It’s really okay if we consider that the library is still `0.x` anyway, so we can clean up all deprecations once we attain `1.0`. However, this leads to the other issues
2. Having a `List<String>` defeats the purpose of having parameters in the first place. Feels like we’re going back to the days of manually parsing path segments and littering everywhere with `pathSegment.get(3)`. Users have to find a disconnected way to keep track of which index references what, or even iterate through the list, both of with is not very optimal. Usually, each group is a distinct value that you’d likely want to treat independently
3. Consider that all parameters will now be `List<String>`, even if it’s just 1 item in each list. Seems quite wasteful, and more so for a library

Reply:
I meant support something like `val params = hashMapOf<String, Any>` and then when a multiple groupings are encountered we do `params[key] = listOf("value1","value2")` Let's keep the current implementation for now and if we have to support multiple groupings, we can refine the current solutions we have available.